cmd/zar/zq/command.go

```diff
@@ -0,0 +1,238 @@
+	err = json.Unmarshal(data, &tc)
+	if err != nil {
+		return nil, fmt.Errorf("%s: unmarshaling error: %s", c.jsonTypePath, err)
+	}
+	if err = tc.Validate(); err != nil {
+		return nil, fmt.Errorf("%s: %s", c.jsonTypePath, err)
+	}
+	return &tc, nil
+}
+
+//XXX lots here copied from zq command... we should refactor into a tools package
+func (c *Command) Run(args []string) error {
+	//XXX
+	if c.outputFile == "-" {
+		c.outputFile = ""
+	}
+
+	if _, err := regexp.Compile(c.jsonPathRegexp); err != nil {
+		return err
+	}
+	if c.jsonTypePath != "" {
+		tc, err := c.loadJsonTypes()
+		if err != nil {
+			return err
+		}
+		c.jsonTypeConfig = tc
+	}
+	if len(args) == 0 {
+		return errors.New("zar zq needs input arguments")
+	}
+	rootPath := c.root
+	if rootPath == "" {
+		rootPath = "."
+	}
+	// XXX this is parallelizable except for writing to stdout when
+	// concatenating results
+	return archive.Walk(rootPath, func(zardir string) error {
+		inputs := args
+		var query ast.Proc
+		var err error
+		first := archive.Localize(zardir, inputs[:1])
+		if first[0] != "" && fileExists(first[0]) {
+			query, err = zql.ParseProc("*")
+			if err != nil {
+				return err
+			}
+		} else {
+			query, err = zql.ParseProc(inputs[0])
+			if err != nil {
+				return err
+			}
+			inputs = inputs[1:]
+		}
+		localPaths := archive.Localize(zardir, inputs)
+		readers, err := c.inputReaders(resolver.NewContext(), localPaths)
+		if err != nil {
+			return err
+		}
+		if len(readers) == 0 {
+			// XXX silently skip if no file found
+			return nil
+		}
+		wch := make(chan string, 5)
+		if !c.stopErr {
+			for i, r := range readers {
+				readers[i] = zbuf.NewWarningReader(r, wch)
+			}
+		}
+		reader := scanner.NewCombiner(readers)
+		defer reader.Close()
+		writer, err := c.openOutput(zardir, c.outputFile)
+		if err != nil {
+			return err
+		}
+		defer writer.Close()
+		// XXX we shouldn't need zap here, nano?  etc
+		mux, err := driver.CompileWarningsCh(context.Background(), query, reader, false, nano.MaxSpan, zap.NewNop(), wch)
+		if err != nil {
+			return err
+		}
+		d := driver.NewCLI(writer)
+		if !c.quiet {
+			d.SetWarningsWriter(os.Stderr)
+		}
+		return driver.Run(mux, d, nil)
+	})
+}
+
+func (c *Command) inputReaders(zctx *resolver.Context, paths []string) ([]zbuf.Reader, error) {
+	cfg := detector.OpenConfig{
+		Format:         c.ReaderFlags.Format,
+		DashStdin:      true,
+		JSONTypeConfig: c.jsonTypeConfig,
+		JSONPathRegex:  c.jsonPathRegexp,
+	}
+	var readers []zbuf.Reader
+	for _, path := range paths {
+		file, err := detector.OpenFile(zctx, path, cfg)
+		if err != nil {
+			// XXX for zar zq siliently conotinue if file doesn't exist
+			if os.IsNotExist(err) {
+				continue
+			}
+			err = fmt.Errorf("%s: %w", path, err)
+			if c.stopErr {
+				return nil, err
+			}
+			fmt.Fprintf(os.Stderr, "%s\n", err)
+			continue
+		}
+		// wrap in a named reader so the reader implements Stringer
+		readers = append(readers, namedReader{file, path})
```

Comment:
I don't see the String() method used, is that for debugging?

Reply:
scanner.Combiner.Read() calls *printf with the reader interface pointer as a %s.   I think Noah came up with this.

Reply:
I will clarify the comment